docs/pallas/tpu/matmul.md

```diff
@@ -0,0 +1,643 @@
+        x_block = x[m_slice, k_slice]
+        y_block = y[k_slice, n_slice]
+        z[m_slice, n_slice] += matmul_small(x_block, y_block)
+  return z
+```
+
++++ {"id": "TP49TV6q8so9"}
+
+Our `block_matmul` function should now work on inputs larger than 256.
+
+```{code-cell} ipython3
+:id: 2SZFnWnurzEC
+
+m, k, n = 4096, 4096, 4096
+x = np.random.uniform(size=(m, k)).astype(np.float32)
+y = np.random.uniform(size=(k, n)).astype(np.float32)
+np.testing.assert_allclose(x @ y, block_matmul(x, y), atol=1e-6, rtol=1e-6)
+```
+
++++ {"id": "GXtjEtEhtARN"}
+
+`block_matmul` decomposes a matrix multiplication into many smaller ones by observing that each output chunk of size `(bm, bn)` can be computed by accumulating several `(bm, bk) x (bk, bn)` size matrix multiplications.
+
+TPUs have native hardware capable of small matrix multiplication akin to `matmul_small`, so to utilize this hardware when doing bigger matrix multiplications, we will apply the `block_matmul` decomposition.
+
++++ {"id": "a0ESFoX1ID0z"}
+
+### Pipelining
+
+In [the previous guide](pipelining), we covered how pipelining in Pallas works. To make sure our compute units are always working and never waiting for memory transfers, we overlap the memory transfers for the next iteration of a kernel with the current one.
+
+In Pallas, we specify that via `BlockSpec`s and a `grid`. Note that we already have a nested for loop in the block matrix multiplication algorithm. We can specify that in Pallas via a `grid`. The slices in the block matrix multiplication can also be specified via `BlockSpec`s.
+
++++ {"id": "FvYoyqlyIqo6"}
+
+## Your first matrix multiplication kernel
+
++++ {"id": "umKZAlSvIt7x"}
+
+Putting it all together, here's an implementation of a block matrix multiplication kernel that pipelines the memory transfers with the compute. We create a 3-d grid, corresponding to the 3-nested loop in the NumPy code. Note that while MXUs are only capable of multiplying small blocks, Pallas will automatically take bigger blocks and automatically tile them over the MXUs.
+
+The last dimension of the grid corresponds to the contraction dimension of the matrix multiply and is a reduction dimension, so we need to be sure to initialize the accumulator.
+
+```{code-cell} ipython3
+:id: 75FBANKFbmQ5
+
+def matmul_kernel(x_ref, y_ref, z_ref):
+  @pl.when(pl.program_id(2) == 0)
+  def _():
+    z_ref[...] = jnp.zeros_like(z_ref)
+
+  z_ref[...] += x_ref[...] @ y_ref[...]
+
+def matmul(
+    x: jax.Array,
+    y: jax.Array,
+    *,
+    bm: int = 128,
+    bk: int = 128,
+    bn: int = 128,
+):
+  m, k = x.shape
+  _, n = y.shape
+  return pl.pallas_call(
+      matmul_kernel,
+      out_shape=jax.ShapeDtypeStruct((m, n), x.dtype),
+      in_specs=[pl.BlockSpec(lambda i, j, k: (i, k), (bm, bk)),
+                pl.BlockSpec(lambda i, j, k: (k, j), (bk, bn))],
+      out_specs=pl.BlockSpec(lambda i, j, k: (i, j), (bm, bn)),
+      grid=(m // bm, n // bn, k // bk)
```

Comment:
I'd like us to make grids parallel on TPU by default, so perhaps let's immediately add dimension semantics annotations